cpu/cortexm_common/include/mpu.h

```diff
@@ -0,0 +1,154 @@
+ *
+ * @}
+ */
+
+#ifndef MPU_H_
+#define MPU_H_
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Number of MPU regions available (will vary depending on the Cortex-M version)
+ */
+#define MPU_NUM_REGIONS ( (MPU->TYPE & MPU_TYPE_DREGION_Msk) >> MPU_TYPE_DREGION_Pos )
+
+/**
+ * @brief Access Permission words
+ */
+enum {
+    AP_NO_NO = 0, /**< no access for all levels */
+    AP_RW_NO = 1, /**< read/write for privileged level, no access from user level */
+    AP_RW_RO = 2, /**< read/write for privileged level, read-only for user level */
+    AP_RW_RW = 3, /**< read/write for all levels */
+    AP_RO_NO = 5, /**< read-only for privileged level, no access from user level */
+    AP_RO_RO = 6, /**< read-only for all levels */
+};
+
+/**
+ * @brief MPU region sizes
+ */
+enum {
+    MPU_SIZE_32B  =  4, /**<  32 bytes     */
+    MPU_SIZE_64B  =  5, /**<  64 bytes     */
+    MPU_SIZE_128B =  6, /**< 128 bytes     */
+    MPU_SIZE_256B =  7, /**< 256 bytes     */
+    MPU_SIZE_512B =  8, /**< 512 bytes     */
+    MPU_SIZE_1K   =  9, /**<   1 kilobytes */
+    MPU_SIZE_2K   = 10, /**<   2 kilobytes */
+    MPU_SIZE_4K   = 11, /**<   4 kilobytes */
+    MPU_SIZE_8K   = 12, /**<   8 kilobytes */
+    MPU_SIZE_16K  = 13, /**<  16 kilobytes */
+    MPU_SIZE_32K  = 14, /**<  32 kilobytes */
+    MPU_SIZE_64K  = 15, /**<  64 kilobytes */
+    MPU_SIZE_128K = 16, /**< 128 kilobytes */
+    MPU_SIZE_256K = 17, /**< 256 kilobytes */
+    MPU_SIZE_512K = 18, /**< 512 kilobytes */
+    MPU_SIZE_1M   = 19, /**<   1 megabytes */
+    MPU_SIZE_2M   = 20, /**<   2 megabytes */
+    MPU_SIZE_4M   = 21, /**<   4 megabytes */
+    MPU_SIZE_8M   = 22, /**<   8 megabytes */
+    MPU_SIZE_16M  = 23, /**<  16 megabytes */
+    MPU_SIZE_32M  = 24, /**<  32 megabytes */
+    MPU_SIZE_64M  = 25, /**<  64 megabytes */
+    MPU_SIZE_128M = 26, /**< 128 megabytes */
+    MPU_SIZE_256M = 27, /**< 256 megabytes */
+    MPU_SIZE_512M = 28, /**< 512 megabytes */
+    MPU_SIZE_1G   = 29, /**<   1 gigabytes */
+    MPU_SIZE_2G   = 30, /**<   2 gigabytes */
+    MPU_SIZE_4G   = 31, /**<   4 gigabytes */
+};
+
+/**
+ * @brief convert a region size code to a size in bytes
+ *
+ * @param[in]   size    region size code, e.g. MPU_SIZE_32B
+ *
+ * @return region size in bytes
+ */
+#define MPU_SIZE_TO_BYTES(size) ( (uintptr_t)1 << ((size) + 1) )
+
+/**
+ * @brief generate an MPU attribute word suitable for writing to the RASR register
+ *
+ * @param[in]   xn      eXecute Never flag (forbids instruction fetches)
+ * @param[in]   ap      Access Permission word, e.g. AP_RO_RO
+ * @param[in]   tex     Type Extension Field
+ * @param[in]   c       Cacheable bit
+ * @param[in]   b       Bufferable bit
+ * @param[in]   s       Sub-Region Disable (SRD) field
+ * @param[in]   size    region size code, e.g. MPU_SIZE_32B
+ *
+ * @return combined region attribute word
+ */
+#define MPU_ATTR(xn, ap, tex, c, b, s, size) ( \
```

Comment:
I prefer the macro but yield to convention. They are binary equivalent for `tests/mpu_stack_guard`.